File: clojure/proverb/src/proverb.clj
(ns proverb
  (:require [clojure.string :as string]))

(defn- line [[want lost]]
  (str "For want of a " want " the " lost " was lost."))

(defn- last-line [want]
  (str "And all for the want of a " want "."))

(defn recite [[first-piece :as pieces]]
  (if-not first-piece
    ""
    (as-> pieces $
          (partition 2 1 $)
          (mapv line $)
          (conj $ (last-line first-piece))
          (string/join "\n" $))))

File: javascript/proverb/proverb.js
export function proverb(...things) {
    let qualifiedThing = things[0];
    if (typeof things[things.length - 1] === 'object') {
        qualifiedThing = `${things[things.length - 1].qualifier} ${things[0]}`;
        things = things.slice(0, things.length - 1);
    }
    return [...things.keys()].slice(1)
                             .map((i) => `For want of a ${things[i - 1]} the ${things[i]} was lost.\n`)
                             .join('') + `And all for the want of a ${qualifiedThing}.`;
}
